src/v/cloud_storage/partition_manifest.cc

```diff
@@ -837,6 +837,29 @@ bool partition_manifest::delete_permanently(
     return false;
 }
 
+partition_manifest::const_iterator
+partition_manifest::segment_containing(model::offset o) const {
+    auto search_key = key{.base_offset = o, .term = model::term_id{0}};
+    auto it = _segments.lower_bound(search_key);
+
+    if (it != _segments.end() && it->first.base_offset == o) {
+        return it;
+    }
+
+    if (it == _segments.begin()) {
+        return end();
+    }
+
+    it = std::prev(it);
```

Comment:
Consider segments starting at 0, 10, 20.

lower_bound(9) -> hit segment 10, seek back 1 (this path)
lower_bound(11) -> hit segment 20, seek back 1 (this path)
lower_bound(10) -> hit segment 10, don't seek (the `return it` path above)

However, this code might be a bit simpler if it used `upper_bound` instead, and does the same thing as `partition_record_batch_reader_impl::seek_segment` (then there's no special case for base_offset==0)

Reply:
the `upper_bound` approach sounds good, I will try to incorporate it.

Reply:
It seems there is a bug in the current code due to using `model::term_id{0}` as the key. This will fail with, for example `m.segment_containing(model::offset{10})` returning `m.end()` rather than the first segment, if the manifest contains offset 10 with term 1.  

This code should use `term_id::max()` to make sure that the offset is the only significant part in calculating upper_bound.

Reply:
added a fix for this and some tests for the new partition manifest method.